### ISSUE TITLE
Performing sequential tests (explicitly not concurrent/parallel)

### ISSUE THREAD

**Issue by roelvandijk** [batterseapower/test-framework#3]
Problem in 1 sentence: My tests fail when performed concurrently.

I have written an interface to a micro-controller which communicates over USB. Since USB connections are a scarce resource (and micro-controllers even more so) I can only open 1 connection at a time. My test suite consists of tests which each open a connection with the 1 micro-controller and then perform some test. The tests fail if they can't open a connection. Only 1 thread/process can have the connection at any single point in time. Therefore any test executed while another is running will fail.

I have "solved" the problem by compiling the test-suite without the "-N" RTS option, but that is not really a satisfactory solution.

A solution would be a special test group, or a parameter for test groups, in which tests are performed sequentially.

**Comment by batterseapower** [comment]
Can you have the tests acquire a lock (i.e. MVar) that protects the shared resource?

**Comment by roelvandijk** [comment]
That is possible, but would require a lot of plumbing. Still, it is a better workaround than my current set of compiler options.

Now, if you would put that MVar inside of some kind of TestGroup :-)

**Comment by batterseapower** [comment]
I was thinking something like this (untested):

```
data MutuallyExcluded t = ME (MVar ()) t

instance Testlike i r t => Testlike i r (MutuallyExcluded t) where
    runTest cto (ME mvar x) = withMVar mvar $ \() -> runTest cto x
    testTypeName ~(ME _ x) = testTypeName x

{-# NOINLINE mutuallyExclusive #-}
mutuallyExclusive :: Test -> Test
mutuallyExclusive = go
  where
    mvar = unsafePerformIO $ newMVar ()

    go (Test tn t)            = Test tn (MutuallyExcluded mvar t)
    go (TestGroup tn ts)      = TestGroup tn (go ts)
    go (PlusTestOptions to t) = PlusTestOptions to (go t)
```

The unsafePerformIO is a bit ugly, I admit

**Comment by batterseapower** [comment]
I uploaded v0.4.0 which includes a mutuallyExclusive combinator like the one above. To make it more robust, I added a way that the Test can include a suspended (IO Test) action that is executed once when preparing the tests to run.

**Comment by roelvandijk** [comment]
I tested the `mutuallyExclusive` combinator and unfortunately it doesn't seem to work for my specific scenario.

I changed my main test function to:

``` haskell
main :: IO ()
main = defaultMain [mutuallyExclusive MB.test]
```

Running this suite will result in most tests failing because the USB connection is busy. Compiling the test suite without `-threaded -rtsopts -with-rtsopts=-N` solves the issue.

Further investigation revealed that the all invocations of the `runTest` method of the `MutuallyExcluded` instance for `TestLike` run in the same thread. I suspect that `runTest` itself spawns another thread, but I don't quite understand the mechanism with the `:~>` constructor. It is also possible that the threads are spawned by QuickCheck.

I would appreciate some advice on what to investigate next. I cloned test-framework locally so I can easily test things.
